File: Python/single_threaded.py
from typing import Final, Set

from web_crawl_utils import fetch_and_parse


class SingleThreadedCrawler:
    def __init__(self, start_url: str, max_depth: int, max_urls: int) -> None:
        self.start_url: Final[str] = start_url
        self.max_depth: Final[int] = max_depth
        self.max_urls: Final[int] = max_urls
        self.visited: Final[Set[str]] = set()

    def start_crawling(self) -> None:
        fetch_and_parse(self.start_url, 0, self.max_depth,
                        self.max_urls, self.visited, None, None)
        print(f"Total URLs crawled (ST): {len(self.visited)}")

File: Go/single_threaded.go
package main

import (
	"context"
	"fmt"
)

var urlCount int = 0

func SequentialCrawl(url string) {
	fetchAndParse(url, 0)
	fmt.Printf("Sequential Crawling finished. Total URLs crawled: %d\n", urlCount)
}

func fetchAndParse(url string, depth int) {
	if depth > maxDepth || urlCount >= maxUrls {
		return
	}
	urlCount++
	// fmt.Printf("Visiting URL #%d: %s at depth: %d\n", urlCount, url, depth)
	if urlCount >= maxUrls {
		return
	}
	FetchAndParseURL(context.TODO(), url, depth, maxDepth, fetchAndParse)
}
